### PR TITLE
Animation: initial mechanism

### PR DESCRIPTION
**Pull request by pkulev**
**Closes issue(s):** [ #26 ]

**Description of the changes:**

  * What's new: Animation system
  * Fixes:
  * Refactoring:

**Reviewers:** [ @taptap ]

**Task list:**

- [x] submit pull request
- [ ] review
- [ ] testing
- [ ] fixes after review, squashing, rebasing to master

### CODE REVIEW

File: xoinvader/animation.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    :param bool loop: loop animation or not
+    """
+
+    def __init__(self, name, bind, attr, keyframes, loop=False):
+        self._name = name
+        self._obj = bind
+        self._attr = attr
+        self._keyframes = sorted(keyframes, key=itemgetter(0))
+        self._loop = loop
+        if not keyframes:
```
Comment (taptap) [inline xoinvader/animation.py]: Why not move this check *before* we init `self._keyframes`?

File: xoinvader/animation.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        if len(self._keyframes) == self._current:
+            if self._loop:
+                self._current = 0
+                self._timer.stop()
+            else:
+                self._timer.stop()
+                raise StopIteration
+
+        keyframe = self._keyframes[self._current]
+        if self._timer.get_elapsed() >= keyframe[0]:
```
Comment (taptap) [inline xoinvader/animation.py]: If would be really cool if the commit message, or method docstring, or preferrably both, contained a brief explanation of the algorithm and it's connection to other engine mechanics.

File: xoinvader/animation.py (hunk 1 above)
Comment (pkulev) [inline xoinvader/animation.py]: Attributes in order. Placing check before makes it worse, I think.

File: xoinvader/animation.py (hunk 2 above)
Comment (pkulev) [inline xoinvader/animation.py]: Will be, when all will be done. Now we just pick next keyframe when previous expired. No interpolation, no complicated logic.

File: xoinvader/animation.py (hunk 1 above)
Comment (taptap) [inline xoinvader/animation.py]: Why?

Comment (pkulev) [inline xoinvader/animation.py]: Visually. I've placed under if code to make noop looping animation if no keyframes given, then deleted it. When we will have an editor, how instantiating will work?

File: xoinvader/animation.py (hunk 2 above)
Comment (taptap) [inline xoinvader/animation.py]: Ok then!

File: xoinvader/ship.py [hunk 3]
```diff
@@ -255,6 +266,22 @@ def __init__(self, pos, border, settings):
         self._fire = True
         self._wbay = Point(x=self._image.width // 2, y=1)
 
+        from xoinvader.animation import AnimationManager
+        self._animgr = AnimationManager()
+        self._animgr.add(
+            name="chaotic", bind=self, attr="direction",
```
Comment (taptap) [inline xoinvader/ship.py]: What is `direction`?
Animation is usually dealing with coordinates, not their deltas.
I think the feature would benefit from example on this too.

Comment (pkulev) [inline xoinvader/ship.py]: This is not related to mechanism itself. Animation binds to attribute in broad sense, not to `coordinate`. We have no sane component to move entities easily, this will be rewritted with new object system.
Direction is dirty thing now, but this is not delta.

File: xoinvader/tests/test_animation.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+import pytest
+
+from xoinvader.animation import Animation, AnimationManager
+
+
+class GameObject(object):
+    """Game object mock."""
+
+    anim_props = dict(
```
Comment (pkulev) [inline xoinvader/tests/test_animation.py]: I think so. Not as common as using dict syntax but valid too.

File: xoinvader/animation.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        self._attr = attr
+        self._keyframes = sorted(keyframes, key=itemgetter(0))
+        self._loop = loop
+        if not keyframes:
+            raise ValueError("Animation keyframes must not be empty.")
+
+        self._timer = Timer(self._keyframes[-1][0], lambda: True)
+        self._timer.start()
+
+        self._current = 0
```
Comment (pkulev) [inline xoinvader/animation.py]: Oh, I think this is obvious. But I can add `# Current keyframe index` comment.

File: xoinvader/animation.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    :param str name: animation name
+    :param object bind: object to bind animation
+    :param str attr: attribute to change in frames
+    :param list keyframes: (float, object) tuples
+    :param bool loop: loop animation or not
+    """
+
+    def __init__(self, name, bind, attr, keyframes, loop=False):
+        self._name = name
```
Comment (pkulev) [inline xoinvader/animation.py]: Ok

File: xoinvader/animation.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+    def add(self, name, *args, **kwargs):
+        """Add new animation, pass args to Animation class.
+
+        :param str name: animation name
+        """
+
+        self._animations[name] = Animation(name, *args, **kwargs)
+        if len(self._animations) == 1:
+            self._current = name
```
Comment (pkulev) [inline xoinvader/animation.py]: mapping {name -> animation object}. Switching of current animation will be implemented through changing self._current. Object that holds AnimationManager knows about animation name that it creates via manager, not about <animation.Animation object at 0x...> =)

File: xoinvader/animation.py (hunk 1 above)
Comment (pkulev) [inline xoinvader/animation.py]: Oh, quite complicated.
You create animation for object.
Code starts work in animation player, but no keyframes created yet.

File: xoinvader/animation.py (hunk 7 above)
Comment (taptap) [inline xoinvader/animation.py]: >Object that holds AnimationManager knows about animation name that it creates via manager, not about <animation.Animation object at 0x...> =)

Yes, but we can traverse `_animations` list and find animation by name when changing it, rather then traverse the dict every time we need access current animation object.

File: xoinvader/ship.py (hunk 3 above)
Comment (taptap) [inline xoinvader/ship.py]: Broader the sense, more complex the interpolation. And we absolutely need it for coordinates.

I think it would be wise to consider coordinates as the main animation target.

File: xoinvader/animation.py (hunk 7 above)
Comment (pkulev) [inline xoinvader/animation.py]: Python not traverses dict. Dict is hashmap, lookup in it is extremely fast.
I think that both approaches are equivalent by time, is this change really needed?
This adds function like `find_animation` with for in it.

File: xoinvader/ship.py (hunk 3 above)
Comment (pkulev) [inline xoinvader/ship.py]: `direction` by sense is Vector2 or Vector3, by implementation now this is an integer due to old implementation. What changes do you want I provide in this PR?

File: xoinvader/animation.py (hunk 5 above)
Comment (pkulev) [inline xoinvader/animation.py]: "obvius" when you looked at update method, of course... Comment is nescessary then, yes.

File: xoinvader/animation.py (hunk 1 above)
Comment (taptap) [inline xoinvader/animation.py]: ...So it immediately rises.
I just don't get it, why we need to call `sorted` on maybe empty list before performing emptiness check.

Comment (pkulev) [inline xoinvader/animation.py]: check on emptiness is in sorting function)

File: xoinvader/animation.py (hunk 7 above)
Comment (taptap) [inline xoinvader/animation.py]: Either this or change `_current` to `_current_name`, for example.
Also, there's no need in making `_animations` list and traversing it in something like `find_animation` list.
You may as well leave it as is, and on update just write `self._current = self._animations[name]`.

Also with this approach things like `self._animations[self._current].update()` will become much more pretty-looking, and things like `self._animations[name].bind(obj)` will remain unchanged.

File: xoinvader/ship.py (hunk 3 above)
Comment (taptap) [inline xoinvader/ship.py]: Well, ok. You're right.
Let's advance incrementally.

I can't wait to look at what `EnemyWave` will look like c:

File: xoinvader/animation.py (hunk 1 above)
Comment (taptap) [inline xoinvader/animation.py]: Okay, but why call this at all. It just illogical.

File: xoinvader/animation.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            raise AttributeError("There is no available animation.")
+
+    @animation.setter
+    def animation(self, name):
+        if name in self._animations:
+            self._animation = self._animations[name]
+        else:
+            raise ValueError("No such animation: '{0}'.".format(name))
+
+    def add(self, name, *args, **kwargs):
```
Comment (taptap) [inline xoinvader/animation.py]: Am I right that `Animation` objects are not intended to be instantiated by hand?

File: xoinvader/animation.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        """Add new animation, pass args to Animation class.
+
+        :param str name: animation name
+        """
+
+        animation = Animation(name, *args, **kwargs)
+        self._animations[name] = animation
+
+        if len(self._animations) == 1:
```
Comment (taptap) [inline xoinvader/animation.py]: I think check like `if not self._animation` would be more appropriate. For example, when we start support removing animations from the manager.

File: xoinvader/tests/test_animation.py (hunk 4 above)
Comment (taptap) [inline xoinvader/tests/test_animation.py]: Ok.

File: xoinvader/tests/test_animation.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+    def keyframe_value(idx):
+        """Return value by idx."""
+        return GameObject.anim_props["keyframes"][idx][1]
+
+    with pytest.raises(ValueError):
+        Animation(bind=None, keyframes=[], **GameObject.dummy)
+
+    obj = GameObject()
+    anim = Animation(bind=obj, **obj.anim_props)
```
Comment (taptap) [inline xoinvader/tests/test_animation.py]: This doesn't look like real-world scenario.
Why don't just create animation keyframe list in place and send it to Animation object?
It will make the test way more easy to understand.

File: xoinvader/animation.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        self._attr = attr
+
+        if not keyframes:
+            raise ValueError("Animation keyframes must not be empty.")
+        self._keyframes = sorted(keyframes, key=itemgetter(0))
+
+        self._loop = loop
+
+        # Timer for tracking local time
+        self._timer = Timer(self._keyframes[-1][0], lambda: True)
```
Comment (taptap) [inline xoinvader/animation.py]: Wait. Does `Timer` always measure real-world time and not some in-game ticks?

File: xoinvader/tests/test_animation.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+    obj.anim = anim
+    assert anim._obj is obj
+    assert anim.name == obj.anim_props["name"]
+
+    anim.update()
+    assert obj.attr == keyframe_value(0)
+    time.sleep(1)
```
Comment (taptap) [inline xoinvader/tests/test_animation.py]: Why real sleep is needed here?

File: xoinvader/tests/test_animation.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    assert animgr.animation == obj.dummy["name"]
+
+    with pytest.raises(ValueError):
+        animgr.animation = "bad-animation"
+
+    animgr.animation = obj.anim_props["name"]
+    assert animgr.animation == obj.anim_props["name"]
+    animgr.animation = obj.dummy["name"]
+
+    assert animgr._animations[obj.dummy["name"]]._obj is obj
```
Comment (taptap) [inline xoinvader/tests/test_animation.py]: Looks like check on Python dictionary, not actual code.

File: xoinvader/animation.py (hunk 8 above)
Comment (taptap) [inline xoinvader/animation.py]: If so, then `*args, **kwargs` is bad, because it doesn't provide any useful information about the interface.

File: xoinvader/animation.py (hunk 11 above)
Comment (pkulev) [inline xoinvader/animation.py]: Why not consider to timer code? :)
Timer is like real timer. Fires after time is exceded.

File: xoinvader/animation.py (hunk 8 above)
Comment (pkulev) [inline xoinvader/animation.py]: Yep, my mistake.

File: xoinvader/animation.py (hunk 11 above)
Comment (taptap) [inline xoinvader/animation.py]: It's quite bad.
How we will stop the time? In paused state for example.
It's not the scope of the current issue however, but I think it would be nicer to implement somthing like is currently managing background advance.

File: xoinvader/tests/test_animation.py (hunk 12 above)
Comment (taptap) [inline xoinvader/tests/test_animation.py]: Ok, got it. Because we indeed measuring real time.

File: xoinvader/animation.py (hunk 9 above)
Comment (pkulev) [inline xoinvader/animation.py]: I think way of this check doesn't matter in both cases.

* we removed all animations
* adding new first animation-> use it because we have single animation

Comment (pkulev) [inline xoinvader/animation.py]: And such code used in state.State class.

File: xoinvader/tests/test_animation.py (hunk 12 above)
Comment (pkulev) [inline xoinvader/tests/test_animation.py]: To do well check of realtime changing frames.

File: xoinvader/animation.py (hunk 9 above)
Comment (taptap) [inline xoinvader/animation.py]: Anyway, it's plain illogical and adds unnecessary algorithmical complexity.
We actually want to know if `self._animation` is already set, and set it if it's not.
So why check for dictionary lenght?

File: xoinvader/tests/test_animation.py (hunk 10 above)
Comment (pkulev) [inline xoinvader/tests/test_animation.py]: This intended to be universal set of arguments, but turns out with unreadable mess, rewriting it.

File: xoinvader/animation.py (hunk 11 above)
Comment (pkulev) [inline xoinvader/animation.py]: Create issue about this please. Really can cause many problems.

File: xoinvader/animation.py (hunk 9 above)
Comment (pkulev) [inline xoinvader/animation.py]: Now agreed.

File: xoinvader/animation.py (hunk 11 above)
Comment (taptap) [inline xoinvader/animation.py]: Ok. Other that that: this will do for now.

File: xoinvader/animation.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    """
+
+    def __init__(self, name, bind, attr, keyframes, loop=False):
+        self._name = name
+        self._obj = bind
+        self._attr = attr
+
+        if not keyframes:
+            raise ValueError("Animation keyframes must not be empty.")
+        self._keyframes = sorted(keyframes, key=itemgetter(0))
```
Comment (taptap) [inline xoinvader/animation.py]: What is `itemgetter(0)`?

File: xoinvader/tests/test_animation.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    with pytest.raises(AttributeError):
+        animgr.animation
+
+    animgr.add("test1", bind=obj, attr="attr", keyframes=[(0.0, 1)])
+    animgr.add("test2", bind=obj, attr="attr", keyframes=[(0.0, 2)])
+
+    assert animgr.animation == "test1"
+
+    with pytest.raises(ValueError):
+        animgr.animation = "bad-animation"
```
Comment (taptap) [inline xoinvader/tests/test_animation.py]: To be honest, `set_animation` function would be better in this case. Assignment would play nice with future CORP-style architecture. Like, 
```
o = GameObject()
o.animation = "swoop-da-whoop"
```
or something like this.

File: xoinvader/animation.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+        self._animations[name] = animation
+
+        if not self._animation:
+            self._animation = animation
+
+    def update(self):
+        """Update manager's state."""
+
+        try:
+            self._animation.update()
```
Comment (taptap) [inline xoinvader/animation.py]: It just occured to me that _every_ object has to have it's own animation manager, since it doesn't update all animations (which may bound to different objects), and can't update animations to same object at once. So one object must have two animation managers for glowing engines and for chaotic movement.

I think we just might get rid of `_animation` field and call update on a whole list.

File: xoinvader/animation.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+class AnimationManager(object):
+    """Manage list of object animation."""
+
+    def __init__(self):
+        self._animations = {}
+        self._animation = None
+
+    @property
+    def animation(self):
+        """AnimationManager's current animation.
```
Comment (pkulev) [inline xoinvader/animation.py]: You changing animation, not name, but getting name. Provide please comment you think fit to this.

File: xoinvader/animation.py (hunk 14 above)
Comment (pkulev) [inline xoinvader/animation.py]: https://docs.python.org/3/library/operator

File: xoinvader/tests/test_animation.py (hunk 15 above)
Comment (pkulev) [inline xoinvader/tests/test_animation.py]: No differense absolutely. First - setter, second - setter too.

File: xoinvader/animation.py (hunk 16 above)
Comment (pkulev) [inline xoinvader/animation.py]: Whoops. Yep. Really. Do we need some sort of states for animation? Active/Inactive. If we are updating all animations then no need to keep current and change them.

File: xoinvader/tests/test_animation.py (hunk 15 above)
Comment (taptap) [inline xoinvader/tests/test_animation.py]: It is, logical one.
AnimationManager changes animation for object.
Assigning animation to object is just like changing an attribute.

File: xoinvader/animation.py (hunk 14 above)
Comment (taptap) [inline xoinvader/animation.py]: Oh! Cool! So it might have been `lambda n: n[0]` instead?

File: xoinvader/animation.py (hunk 16 above)
Comment (taptap) [inline xoinvader/animation.py]: I think it would be nice.

File: xoinvader/animation.py (hunk 17 above)
Comment (taptap) [inline xoinvader/animation.py]: Yes, you're right. 

Hmmmm... I think this peculiarity can be described in the docstring, and title may remain unchanged then.

File: xoinvader/tests/test_animation.py (hunk 15 above)
Comment (pkulev) [inline xoinvader/tests/test_animation.py]: I think that future is most appropriate time for discussing style of attribute setters. Now this done according to other code. When we construct good object system that will glue such components - then this conversation will have sense.

File: xoinvader/animation.py (hunk 14 above)
Comment (pkulev) [inline xoinvader/animation.py]: This was invented to use instead of lambda item: item[any_number].

File: xoinvader/tests/test_animation.py (hunk 15 above)
Comment (taptap) [inline xoinvader/tests/test_animation.py]: I agree.

File: xoinvader/animation.py (hunk 14 above)
Comment (taptap) [inline xoinvader/animation.py]: Understood. Thanks.

**Comment by taptap** [comment]
>**feature-behaviour**
Peculiar branch name. Just noticed.

File: xoinvader/animation.py (hunk 17 above)
Comment (pkulev) [inline xoinvader/animation.py]: LOLO, nice feature. Fire rocket, go to menu, wait, go back in game. There are two rockets already fly towards the world's end.

Comment (taptap) [inline xoinvader/animation.py]: My high score is 4 rockets.

Why this happens?

File: xoinvader/animation.py (hunk 16 above)
Comment (taptap) [inline xoinvader/animation.py]: On the second thought, it may be premature. Let's postpone this for later.
